design/local-node-agent.md

```diff
@@ -0,0 +1,123 @@
+The Rook operator will deploy the Rook agent to all nodes in the cluster via a Daemonset in the same namespace in which the operator is running.
+It is a permanent (long running) daemon that has a lifecycle tied to the node that it is scheduled on.
+The agent deployment will happen when the operator is first created, in the same flow where the operator is declaring CRDs for clusters, pools, etc.
+This means that the Rook agents are not associated with a specific Rook cluster and that they will be able to handle operations for any Rook cluster instance.
+The Cluster CRD will be updated to allow selectors to control the set of nodes that the agent is scheduled on, but in most cases it is desirable for it to be running on all nodes.
+
+Each agent pod will be running the same `rook/rook` container image in use today, but with a new `agent` command (similar to the existing `mon` or `osd` commands).
+When the agent starts up, it will immediately copy the Rook Flexvolume driver from its container image to the `volume-plugin-dir` on its host node.
+In Kubernetes 1.8+, the [dynamic Flexvolume plugin discovery](https://github.com/kubernetes/community/pull/833) will find and initialize our driver, but in older versions of Kubernetes a manual restart of the Kubelet will be required.
+
+After the driver has been deployed, the Rook agent will create a Unix domain socket that will serve as the communications channel between the agent and the Flexvolume driver.
+The agent will wait for the driver to complete a handshake over the socket before moving on and being ready to accept Flexvolume storage requests.
+
+Note that the Rook operator will continue to run its existing [dynamic volume provisioner](https://github.com/rook/rook/tree/master/pkg/operator/provisioner) to provision and delete persistent volumes as needed.
+
+### **Flexvolume**
+#### **Control Flow Overview**
+The Rook Flexvolume driver will be very lightweight, simply implementing the [required interface](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md#driver-invocation-model)
+and then offloading the storage provider work to the Rook agent over the Unix domain socket.
+Below is a simplified description of the control flow for providing block storage ready to be consumed by a pod:
+
+1. Rook operator is running and a Rook agent pod is running on every node in the cluster.  The Rook Flexvolume driver has been deployed to the `volume-plugin-dir` on each node.
+1. A user creates a pod with a Persistent Volume Claim (PVC) specifying a storageclass that uses the `rook.io/block` [provisioner](https://github.com/kubernetes-incubator/external-storage/blob/e6e64ad1a431fea37f723882f36251f8d2fe4247/lib/controller/volume.go#L29)
+1. `Provision()` is called on the operator's provisioner to create a block image in the cluster
+1. `Attach()` is called by the attach/detach controller on the Rook Flexvolume, which calls into the Rook agent via the Unix domain socket.
```

Comment:
Just to clarify the steps, the attach() flow will start when the user wants to consume the storage. ie user pod that uses that storage gets created.